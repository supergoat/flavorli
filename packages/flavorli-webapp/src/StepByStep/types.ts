import {IColor} from '@flavorli/elements/lib/theme/colors';

export interface ITag {
  text: string;
  color: IColor;
}

export interface ILink {
  heading: string;
  name: string;
  from: number;
  timerId?: number;
}

export interface IIngredient {
  qty: string;
  name: string;
}

export interface ITimer {
  id: number;
  name: string;
  minutes: number;
  seconds: number;
  isPaused?: boolean;
}

export interface IKitchenware {
  qty: string;
  name: string;
}

export interface IImage {
  src: string;
  alt: string;
}
export interface IStep {
  type: 'PREPARATION' | 'MISE_EN_PLACE_STEP';
  no: number;
  description: string;
  tag: ITag;
  links: ILink[];
  ingredients: IIngredient[];
  timer?: ITimer;
  kitchenware: IKitchenware[];
  images?: IImage[];
}

export interface IItem {
  qty: string;
  name: string;
}

export interface IMiseEnPlaceStep {
  no: number;
  type: 'MISE_EN_PLACE';
  heading: string;
  image: string;
  items: IItem[];
}

export interface IIntro {
  type: 'INTRO';
  name: string;
  image: string;
  preparation: string;
  cooking: string;
  portions: string;
  difficulty: string;
}
