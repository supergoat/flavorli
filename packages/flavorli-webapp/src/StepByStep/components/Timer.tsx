import React from 'react';
import {Stack, Text, Button} from '@flavorli/elements';
import {
  useAddRecipeTimerIfItDoesNotExist,
  useRecipeTimer,
  convertMsToMinsAndSecs,
  useToggleTimer,
  useResetTimer,
} from '../useRecipeTimer';
import {ITimer} from '../../types';

interface ITimerProps {
  type?: 'notification';
  timerInfo: ITimer;
}

const Timer = ({timerInfo, type}: ITimerProps) => {
  const timer = useAddRecipeTimerIfItDoesNotExist(timerInfo);

  const remainingTimerInMs = useRecipeTimer(
    timer.updatedAt,
    timer.remainingTime,
    timer.isPaused,
  );

  const {minutes, seconds} = convertMsToMinsAndSecs(remainingTimerInMs);

  const toggleTimer = useToggleTimer(timer, remainingTimerInMs);
  const resetTimer = useResetTimer(timerInfo);

  return (
    <Stack
      width="100%"
      gap={4}
      direction={type === 'notification' ? 'horizontal' : 'vertical'}
      distribution={type === 'notification' ? 'space-between' : 'start'}
      alignment={type === 'notification' ? 'center' : 'start'}
    >
      <Text
        id={`timer-${timer.id}`}
        role="timer"
        aria-label={timer.name}
        aria-atomic={true}
        fontSize={type === 'notification' ? 24 : 32}
      >
        {minutes}m {seconds}s
      </Text>

      <Stack direction="horizontal" gap={8}>
        {(minutes > 0 || seconds > 0) && (
          <Button
            intent="text"
            onClick={toggleTimer}
            aria-controls={`timer-${timer.id}`}
          >
            {timer.isPaused ? 'START' : 'PAUSE'}
          </Button>
        )}
        <Button intent="text" color="secondaryTextColor" onClick={resetTimer}>
          RESET
        </Button>
      </Stack>
    </Stack>
  );
};

export default Timer;
