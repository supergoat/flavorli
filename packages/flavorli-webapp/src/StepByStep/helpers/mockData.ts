import {IStep} from '../types';

export const steps: IStep[] = [
  {
    no: 1,
    tag: {
      text: 'Sauce',
      color: 'tagRed',
    },
    links: [],
    description: 'Place a pot of water in medium heat and set aside to boil',
    ingredients: [],
    kitchenware: [
      {
        qty: '1',
        name: 'Pot',
      },
    ],
  },
  {
    no: 2,
    tag: {
      text: 'Sauce',
      color: 'tagRed',
    },
    links: [],
    description: 'Place a sauce pan over high heat and add the olive oil',
    ingredients: [
      {
        qty: '4 - 6 tbsps',
        name: 'olive oil',
      },
    ],
    kitchenware: [
      {
        qty: '1',
        name: 'Sauce Pan',
      },
    ],
  },
  {
    no: 3,
    tag: {
      text: 'Sauce',
      color: 'tagRed',
    },
    links: [],
    description:
      'Cut the eggplan into 1 cm cubes. Add it to the pan and sauté for 1-2 minutes',
    ingredients: [
      {
        qty: '2',
        name: 'eggplants',
      },
    ],
    kitchenware: [
      {qty: '1', name: 'Knife'},
      {qty: '1', name: 'Cutting Board'},
      {qty: '1', name: 'Spoon'},
    ],
  },
  {
    no: 4,
    tag: {
      text: 'Sauce',
      color: 'tagRed',
    },
    links: [],
    description:
      'Finely chop the onion, add it to the pan and sauté for 1-2 minutes',
    ingredients: [
      {
        qty: '1',
        name: 'onion',
      },
    ],
    kitchenware: [],
  },
  {
    no: 5,
    tag: {
      text: 'Sauce',
      color: 'tagRed',
    },
    links: [],
    description:
      'Cut the zucchinis and the peppers into small cubes and add them to the pan. Sauté for 4-5 minutes, until the vegetables are ready',
    ingredients: [
      {
        qty: '2',
        name: 'zucchinis',
      },
      {
        qty: '1',
        name: 'red bell pepper',
      },
      {
        qty: '1',
        name: 'green bell pepper',
      },
    ],
    kitchenware: [],
  },
  {
    no: 6,
    tag: {
      text: 'Sauce',
      color: 'tagRed',
    },
    links: [],
    description:
      'Add the salt, pepper, oregano, thyme, chili flakes, sugar and tomato paste and sauté',
    ingredients: [
      {
        qty: '',
        name: 'salt',
      },
      {
        qty: '',
        name: 'pepper',
      },
      {
        qty: '1 tsp',
        name: 'oregano',
      },
      {
        qty: '1 tbsp',
        name: 'thyme',
      },
      {
        qty: '1 tsp',
        name: 'chilli flakes',
      },
      {
        qty: '1 tbsp',
        name: 'granulated sugar',
      },
      {
        qty: '1 tsp',
        name: 'tomato paste',
      },
    ],
    kitchenware: [
      {
        qty: '1',
        name: 'Tea Spoon',
      },
    ],
  },
  {
    no: 7,
    tag: {
      text: 'Sauce',
      color: 'tagRed',
    },
    links: [],
    description:
      'Add the wine and canned tomatoes and simmer at medium-low heat for 10 minutes',
    ingredients: [
      {
        qty: '40 g',
        name: 'white wine',
      },
      {
        qty: '400 g',
        name: 'canned tomatoes',
      },
    ],
    timer: {id: 7, name: 'Step 7: Simmer Sauce', minutes: 10, seconds: 0},
    kitchenware: [{qty: '1', name: 'scale'}],
  },
  {
    no: 8,
    tag: {
      text: 'Sauce',
      color: 'tagRed',
    },
    links: [{from: 1, heading: 'From Step 1', name: 'Boiling water'}],
    description:
      'In the pot with boiling water, add salt along with the macaroni, and boil for ten minutes',
    ingredients: [
      {
        qty: '',
        name: 'salt',
      },
      {
        qty: '500 g',
        name: 'macaroni',
      },
    ],
    timer: {id: 8, name: 'Step 8: Cook Pasta', minutes: 10, seconds: 0},
    kitchenware: [],
  },
  {
    no: 9,
    tag: {
      text: 'Assembling',
      color: 'tagPink',
    },
    links: [],
    description: 'Preheat the oven at 180 C set to fan',
    ingredients: [],
    kitchenware: [],
  },
  {
    no: 10,
    tag: {
      text: 'Bechamel',
      color: 'tagOrange',
    },
    links: [],
    description:
      'Place a pot over medium heat. Add the olive oil and the flour and using a hand whisk, mix well until all of the flour is absorbed',
    ingredients: [
      {
        qty: '80 g',
        name: 'olive oil',
      },
      {
        qty: '80 g',
        name: 'all-purpose flour',
      },
    ],
    kitchenware: [
      {qty: '1', name: 'Small Pot'},
      {qty: '1', name: 'Hand Whisk'},
    ],
  },
  {
    no: 11,
    tag: {
      text: 'Bechamel',
      color: 'tagOrange',
    },
    links: [],
    description: 'Add the soy milk in 7-8 batches, stirring constantly',
    ingredients: [
      {
        qty: '800 g',
        name: 'soy milk',
      },
    ],
    kitchenware: [],
  },
  {
    no: 12,
    tag: {
      text: 'Bechamel',
      color: 'tagOrange',
    },
    links: [],
    description:
      'Season with salt and pepper and as soon as it thickens and comes to boil, remove from the heat',
    ingredients: [
      {
        qty: '',
        name: 'salt',
      },
      {
        qty: '',
        name: 'pepper',
      },
    ],
    kitchenware: [],
  },
  {
    no: 13,
    tag: {
      text: 'Sauce',
      color: 'tagRed',
    },
    links: [
      {
        from: 7,
        heading: 'From Step 7',
        name: 'Pan with vegetables',
        timerId: 7,
      },
      {
        from: 8,
        heading: 'From Step 8',
        name: 'Macaroni',
        timerId: 8,
      },
    ],
    description:
      'Drain the macaroni and transfer to the pan with the vegetables. Mix and set aside',
    ingredients: [],
    kitchenware: [{qty: '1', name: 'Colander'}],
  },
  {
    no: 14,
    tag: {
      text: 'Assembling',
      color: 'tagPink',
    },
    links: [{from: 12, heading: 'From Step 12', name: 'Bechamel sauce'}],
    description:
      'Add 100g of the bechamel into the pasta mixture, add oregano, mint, and mix well',
    ingredients: [
      {
        qty: '1 tsp',
        name: 'oregano',
      },
      {
        qty: '1 tbsp',
        name: 'mint',
      },
    ],
    kitchenware: [],
  },
  {
    no: 15,
    tag: {
      text: 'Assembling',
      color: 'tagPink',
    },
    links: [],
    description:
      'Grease 40 cm ovenproof baking dish with olive oil and sprinke the breadcrumps',
    ingredients: [
      {
        qty: '1 tbsp',
        name: 'olive oil',
      },
      {
        qty: '25 g',
        name: 'breadcrumps',
      },
    ],
    kitchenware: [{qty: '1', name: '40 cm ovenproof baking dish'}],
  },
  {
    no: 16,
    tag: {
      text: 'Assembling',
      color: 'tagPink',
    },
    links: [],
    description:
      'Add the pasta mixture, spread the bechamel sauce, and sprinke the breadcrumps, olive oil, thyme and pepper',
    ingredients: [
      {
        qty: '25 g',
        name: 'breadcrumps',
      },
      {
        qty: '1 tbsp',
        name: 'olive oil',
      },
      {
        qty: '1 tbsp',
        name: 'thyme',
      },
      {
        qty: '',
        name: 'pepper',
      },
    ],
    kitchenware: [],
  },
  {
    no: 17,
    tag: {
      text: 'Assembling',
      color: 'tagPink',
    },
    links: [],
    description: 'Bake for 25 - 30 minutes. Remove and let it cool',
    ingredients: [],
    timer: {id: 17, name: 'Step 17: Bake Pastitsio', minutes: 30, seconds: 0},
    kitchenware: [
      {
        qty: '1',
        name: 'Oven Mitts',
      },
    ],
  },
];
