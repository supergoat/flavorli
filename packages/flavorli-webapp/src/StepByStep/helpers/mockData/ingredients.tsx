export const ingredients = [
  {
    qty: '2',
    name: 'eggplants',
  },
  {
    qty: '1',
    name: 'onion',
  },
  {
    qty: '1 clove',
    name: 'garlic',
  },
  {
    qty: '2',
    name: 'zucchinis',
  },
  {
    qty: '1',
    name: 'red bell pepper',
  },
  {
    qty: '1',
    name: 'green bell pepper',
  },
  {
    qty: '2 teaspoon',
    name: 'oregano',
  },
  {
    qty: '2 tablespoon',
    name: 'thyme',
  },
  {
    qty: '1 teaspoon',
    name: 'chilli flakes',
  },
  {
    qty: '1 tablespoon',
    name: 'granulated sugar',
  },
  {
    qty: '1 tablespoon',
    name: 'tomato paste',
  },
  {
    qty: '50 g',
    name: 'white wine',
  },
  {
    qty: '400 g',
    name: 'canned tomatoes',
  },
  {
    qty: '500 g',
    name: 'macaroni',
  },
  {
    qty: '80 g',
    name: 'all-purpose flour',
  },
  {
    qty: '80 g',
    name: 'olive oil',
  },
  {
    qty: '800 g',
    name: 'soy milk',
  },
  {
    qty: '50 g',
    name: 'dry breadcrumps',
  },
  {
    qty: '1 tablespoon',
    name: 'mint',
  },
  {
    qty: '1',
    name: 'squeeze bottle olive oil',
  },
  {
    qty: '',
    name: 'salt',
  },
  {
    qty: '',
    name: 'pepper',
  },
];
