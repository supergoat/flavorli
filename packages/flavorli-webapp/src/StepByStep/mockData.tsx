import {IStep, IRecipeTimer, ILink, IRecipe} from '../types';

export const recipes: IRecipe[] = [
  {
    id: '1',
    createdAt: '2020-01-19T19:37:42.873Z',
    author: 'Robin Robertson',
    name: 'Cherry Berry Smoothie',
    image:
      'https://flavorli-bucket-2.s3.eu-west-2.amazonaws.com/CherryBerrySmoothie.jpeg',
    preparationTime: 7,
    cookingTime: 0,
    portions: '1',
    difficulty: '1',
    ingredients: [
      {
        qty: '1 cup/100 g',
        name: 'Frozen Blueberries',
      },
      {
        qty: '1/2 cup/110 g',
        name: 'fresh or frozen pitted cherries',
      },
      {
        qty: '1',
        name: 'frozen ripe banana, cut into chunks before freezing',
      },
      {
        qty: '1 tbsp',
        name: 'ground flaxseeds (or linseeds)',
      },
      {
        qty: '1 tbsp',
        name: 'almond butter (or peanut butter)',
      },
      {
        qty: '1 cup/22.5 g',
        name: 'raw spinach leaves',
      },
      {
        qty: '1 1/2 cup (360 ml)',
        name: 'water',
      },
    ],
    items: [
      {
        qty: '1',
        name: 'Blender',
      },
      {
        qty: '1',
        name: 'Cup',
      },
      {
        qty: '1/2',
        name: 'cup',
      },
      {
        qty: '1',
        name: 'tablespoon',
      },
    ],
    steps: [
      {
        no: 1,
        type: 'MISE_EN_PLACE',
        tag: {
          text: 'Smoothie',
          color: 'tagRed',
        },
        items: [
          {
            qty: '1',
            name: 'Blender',
          },
        ],
        ingredients: [
          {
            qty: '1 cup/100 g',
            name: 'Frozen Blueberries',
          },
          {
            qty: '1/2 cup/110 g',
            name: 'fresh or frozen pitted cherries',
          },
          {
            qty: '1',
            name: 'frozen ripe banana, cut into chunks before freezing',
          },
          {
            qty: '1 tbsp',
            name: 'ground flaxseeds (or linseeds)',
          },
          {
            qty: '1 tbsp',
            name: 'almond butter (or peanut butter)',
          },
          {
            qty: '1 cup/22.5 g',
            name: 'raw spinach leaves',
          },
          {
            qty: '1 1/2 cup (360 ml)',
            name: 'water',
          },
        ],
        tasks: [
          'Add 1 cup/100 g Frozen Blueberries in a blender',
          'Add 1/2 cup/110 g fresh or frozen pitted cherries',
          'Add 1 frozen ripe banana, cut into chunks before freezing',
          'Add 1 tbsp ground flaxseeds (or linseeds)',
          'Add 1 tbsp almond butter (or peanut butter)',
          'Add 1 cup/22.5 g raw spinach leaves',
          'Add 1 1/2 cup cup (360 ml) water',
        ],
        images: [],
      },
      {
        no: 2,
        type: 'PREPARATION',
        tag: {
          text: 'Smoothie',
          color: 'tagRed',
        },
        items: [],
        ingredients: [],
        tasks: [
          'Blend all the ingredients until smooth and creamy for about 1 minute',
          'Add more water if you prefer a thinner consistency',
          'Serve immediately',
        ],
        timer: {
          id: '2',
          name: 'Step 2: Blend',
          minutes: 1,
          seconds: 0,
        },
        images: [],
      },
    ],
  },
];

export const timer: IRecipeTimer = {
  id: '1',
  name: 'Thicken Sauce',
  minutes: 10,
  seconds: 0,
  isPaused: true,
  updatedAt: '2020-01-26T15:15:06.689Z',
  remainingTime: 600000,
};

export const link: ILink = {
  from: 1,
  heading: 'From Step 1',
  name: 'Thicken Sauce',
};

export const linkWithTimer: ILink = {
  from: 1,
  heading: 'From Step 1',
  name: 'Thicken Sauce',
  timerId: '1',
};

export const items = [
  {
    qty: '1',
    name: 'Pot',
  },
];

export const preparationStep: IStep = {
  type: 'MISE_EN_PLACE',
  tag: {
    text: 'Sauce',
    color: 'tagRed',
  },
  tasks: ['Add the coconul oil in a ramekin and place it first in order'],
  ingredients: [
    {
      qty: '2 tbsps',
      name: 'coconut oil',
    },
  ],
  items: [
    {
      qty: '1',
      name: 'Ramekin',
    },
  ],
  images: [],
};

export const recipeStep: IStep = {
  type: 'PREPARATION',
  no: 1,
  tag: {
    text: 'Sauce',
    color: 'tagRed',
  },
  links: [],
  tasks: ['Heat the coconut oil in a small pot on low heat'],
  ingredients: [
    {
      qty: '2 tbsps',
      name: 'coconut oil',
    },
  ],
  items: [
    {
      qty: '1',
      name: 'Small Pot',
    },
  ],
  images: [],
};

export const recipeSteps = [recipeStep];

export const introStep = {
  author: 'Akis Petratzikis',
  name: 'Greek Pastitsio',
  image: '',
  preparation: '25 minutes',
  cooking: '40 minutes',
  portions: '8 - 10',
  difficulty: 'Level 2',
};

export const ingredients = [{qty: '1', name: 'Onion'}];

export const images = [
  {
    src: '',
    alt: 'image 1',
  },

  {
    src: '',
    alt: 'image 2',
  },
];
