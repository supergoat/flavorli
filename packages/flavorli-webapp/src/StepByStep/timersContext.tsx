import React from 'react';
import {IRecipeTimer} from '../types';
import {getSavedTimersFromLocalStorage} from './useRecipeTimer';

interface IRecipeTimersContext {
  recipeId: string;
  recipeTimers: {[timerId: string]: IRecipeTimer};
  setRecipeTimers: React.Dispatch<{
    [timerId: string]: IRecipeTimer;
  }>;
}

export const RecipeTimersContext = React.createContext<IRecipeTimersContext | null>(
  null,
);

export function RecipeTimersProvider({
  recipeId,
  initialValues,
  ...props
}: {
  recipeId: string;
  initialValues?: {[timerId: string]: IRecipeTimer};
  children: React.ReactNode;
}) {
  const [initState] = React.useState(
    initialValues || getSavedTimersFromLocalStorage()[recipeId] || {},
  );

  const [recipeTimers, setRecipeTimers] = React.useReducer(
    (
      state: {[timerId: string]: IRecipeTimer},
      action: {[timerId: string]: IRecipeTimer},
    ) => ({
      ...state,
      ...action,
    }),
    initState,
  );

  return (
    <RecipeTimersContext.Provider
      value={{recipeTimers, setRecipeTimers, recipeId}}
      {...props}
    />
  );
}

export function useRecipeTimersContext() {
  const context = React.useContext(RecipeTimersContext);

  if (!context) {
    throw new Error(
      'useRecipeTimersContext must be used within a TimersProvider',
    );
  }
  return context;
}
