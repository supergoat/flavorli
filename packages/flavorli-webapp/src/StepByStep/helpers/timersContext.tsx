import React from 'react';
import {ITimer} from '../types';
import {useInterval} from '../../helpers/hooks';

interface ITimersContext {
  timers: {[timerId: number]: ITimer};
  setTimers: React.Dispatch<React.SetStateAction<{}>>;
}

export const TimersContext = React.createContext<ITimersContext | null>(null);

export function TimersProvider({
  initialValues,
  ...props
}: {
  initialValues?: Partial<ITimersContext>;
  children: React.ReactNode;
}) {
  const [contextTimers, setTimers] = React.useState({});

  const timers = initialValues?.timers;

  React.useEffect(() => {
    if (timers) setTimers(timers);
  }, [timers]);

  if (!contextTimers) return null;

  return (
    <TimersContext.Provider
      value={{timers: contextTimers, setTimers}}
      {...props}
    />
  );
}

export function useTimersContext() {
  const context = React.useContext(TimersContext);
  if (!context) {
    throw new Error('useTimersContext must be used within a TimersProvider');
  }
  return context;
}

export function useAddTimerIfItDoesNotExist(timer: ITimer) {
  const {timers, setTimers} = useTimersContext();
  if (timers[timer.id]) return;
  setTimers(t => ({...t, [timer.id]: {...timer, isPaused: true}}));
}

export function useUpdateTimers(timer: ITimer) {
  const {setTimers} = useTimersContext();

  useInterval(() => {
    if (timer.isPaused) return;

    if (timer.minutes === 0 && timer.seconds === 0) {
      setTimers(t => ({
        ...t,
        [timer.id]: {...timer, isPaused: true},
      }));
    } else if (timer.seconds > 0) {
      setTimers(t => ({
        ...t,
        [timer.id]: {...timer, seconds: timer.seconds - 1},
      }));
    } else {
      setTimers(t => ({
        ...t,
        [timer.id]: {...timer, minutes: timer.minutes - 1, seconds: 59},
      }));
    }
  }, 1000);
}
