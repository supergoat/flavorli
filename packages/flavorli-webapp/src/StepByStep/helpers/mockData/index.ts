// https://www.youtube.com/watch?v=q1q3HfVvzQ0
import {ingredients} from './ingredients';
import {items} from './items';
import {preparationSteps} from './preparationSteps';
import {recipeSteps} from './recipeSteps';
import {IStepByStep} from '../../types';

export const stepByStep: IStepByStep = {
  intro: {
    name: 'Greek Pastitsio',
    image:
      'https://akispetretzikis.com/system/uploads/medium/data/14955/recipe_main_vegan-pastitsio.jpg',
    preparation: '25 minutes',
    cooking: '40 minutes',
    portions: '8 - 10',
    difficulty: 'Level 2',
  },
  ingredients,
  items,
  preparationSteps,
  recipeSteps,
};

export {ingredients, items, preparationSteps, recipeSteps};
