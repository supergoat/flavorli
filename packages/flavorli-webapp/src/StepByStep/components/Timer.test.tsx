import React from 'react';
import {render} from '../../helpers/test-helpers';
import {axe} from 'jest-axe';
import Timer from './Timer';
import {steps} from '../helpers/mockData';
import {act} from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import {ITimer} from '../types';
import {TimersProvider} from '../helpers/timersContext';

afterEach(() => {
  jest.useRealTimers();
});

const setup = (type?: 'notification') => {
  const stepWithTimer = steps[6];
  const timer = stepWithTimer.timer as ITimer;

  return {
    ...render(
      <TimersProvider
        initialValues={{timers: {[timer.id]: {...timer, isPaused: true}}}}
      >
        <Timer id={timer.id} type={type} />
      </TimersProvider>,
    ),
    timer,
  };
};

describe('Timer', () => {
  it('should not have any axe violations', async () => {
    const {container} = setup();
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });

  it('should have a role of "timer" to inform assistive technology users that it is a timer', () => {
    const {getByRole} = setup();
    getByRole('timer');
  });

  it('should have an aria-label to provide the name of the timer', () => {
    const {getByLabelText, timer} = setup();
    getByLabelText(timer.name);
  });

  it('should have a aria-atomic set to true to ensure the entirety of the time is announced in full to assistive technology users', () => {
    const {getByRole} = setup();
    expect(getByRole('timer')).toHaveAttribute('aria-atomic', 'true');
  });

  it('should have aria-controls pointing to the timer on the start/pause buttons that control the timer', () => {
    const {getByRole, getByText, timer} = setup();
    expect(getByRole('timer')).toHaveAttribute('id', `timer-${timer.id}`);

    const startButton = getByText('START');
    expect(startButton).toHaveAttribute('aria-controls', `timer-${timer.id}`);

    // Start the timer to have access to the pause button
    act(() => {
      userEvent.click(startButton);
    });

    const pauseButton = getByText('PAUSE');
    expect(pauseButton).toHaveAttribute('aria-controls', `timer-${timer.id}`);
  });

  it('should render correctly', () => {
    const {container} = setup();
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render correctly when type is notification', () => {
    const {container} = setup('notification');
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render an empty div if the timer is null', () => {
    const {container} = render(<Timer />);
    expect(container.firstChild).toBeNull();
  });
});
