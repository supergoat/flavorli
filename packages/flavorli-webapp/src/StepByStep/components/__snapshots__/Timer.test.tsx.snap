// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timer should render correctly 1`] = `
<div>
  <div
    class="sc-AykKH lkPDtk"
    direction="vertical"
    width="100%"
  >
    <p
      aria-atomic="true"
      aria-label="Thicken Sauce"
      class="sc-AykKG jBmvtL"
      font-size="32"
      id="timer-1"
      role="timer"
    >
      10
      m 
      0
      s
    </p>
    <div
      class="sc-AykKH cRsFPk"
      direction="horizontal"
    >
      <button
        aria-controls="timer-1"
        class="sc-AykKC hYfwNR"
        width="47px"
      >
        START
      </button>
      <button
        class="sc-AykKC dAUeMx"
        color="secondaryTextColor"
      >
        RESET
      </button>
    </div>
  </div>
</div>
`;

exports[`Timer should render correctly when type is notification 1`] = `
<div>
  <div
    class="sc-AykKH cKRlwi"
    direction="horizontal"
    width="100%"
  >
    <p
      aria-atomic="true"
      aria-label="Thicken Sauce"
      class="sc-AykKG hCeYPE"
      font-size="24"
      id="timer-1"
      role="timer"
    >
      10
      m 
      0
      s
    </p>
    <div
      class="sc-AykKH cRsFPk"
      direction="horizontal"
    >
      <button
        aria-controls="timer-1"
        class="sc-AykKC hYfwNR"
        width="47px"
      >
        START
      </button>
      <button
        class="sc-AykKC dAUeMx"
        color="secondaryTextColor"
      >
        RESET
      </button>
    </div>
  </div>
</div>
`;
